### ISSUE TITLE
[enhancement] Executive::execute_block should return error message

### ISSUE THREAD

**Issue by lamafab** [paritytech/substrate#6716]
Hi

The Runtime entry `Core_execute_block` of Polkadot (which ends up calling `Executive::execute_block`) should return an error message if there's something wrong with execution. Currently the function just panics, which is fine in order to interpret the execution as an error, but it would be useful to have a more specific error message. There is no way to tell **why** the execution failed.

I guess it's up for debate whether knowing why a block failed is even important or useful, but given that the Polkadot Host cannot interpret an opaque blob of data, this might be useful for various cases. Such as monitoring, debugging or other intelligence/analytics. 

***

The error should be similar to `apply_extrinsic`, which returns a precise message of a single extrinsic:

```rust
/// The result of applying of an extrinsic.
///
/// This type is typically used in the context of `BlockBuilder` to signal that the extrinsic
/// in question cannot be included.
///
/// A block containing extrinsics that have a negative inclusion outcome is invalid. A negative
/// result can only occur during the block production, where such extrinsics are detected and
/// removed from the block that is being created and the transaction pool.
///
/// To rehash: every extrinsic in a valid block must return a positive `ApplyExtrinsicResult`.
///
/// Examples of reasons preventing inclusion in a block:
/// - More block weight is required to process the extrinsic than is left in the block being built.
///   This doesn't necessarily mean that the extrinsic is invalid, since it can still be
///   included in the next block if it has enough spare weight available.
/// - The sender doesn't have enough funds to pay the transaction inclusion fee. Including such
///   a transaction in the block doesn't make sense.
/// - The extrinsic supplied a bad signature. This transaction won't become valid ever.
pub type ApplyExtrinsicResult = Result<DispatchOutcome, transaction_validity::TransactionValidityError>
```

The method `execute_extrinsics_with_book_keeping` and `apply_extrinsic_no_note` should be extended to return an error if an individual extrinsic fails.

```rust
	/// Execute given extrinsics and take care of post-extrinsics book-keeping.
	fn execute_extrinsics_with_book_keeping(extrinsics: Vec<Block::Extrinsic>, block_number: NumberFor<Block>) {
		extrinsics.into_iter().for_each(Self::apply_extrinsic_no_note);

		// ...
	}

	/// Apply an extrinsic inside the block execution function.
	fn apply_extrinsic_no_note(uxt: Block::Extrinsic) {
		let l = uxt.encode().len();
		match Self::apply_extrinsic_with_len(uxt, l, None) {
			Ok(_) => (),
			Err(e) => { let err: &'static str = e.into(); panic!(err) },
		}
	}
```

Similar to `apply_extrinsic`, but should also return the erroneous extrinsic itself.

```rust
	pub fn apply_extrinsic(uxt: Block::Extrinsic) -> ApplyExtrinsicResult {
		let encoded = uxt.encode();
		let encoded_len = encoded.len();
		Self::apply_extrinsic_with_len(uxt, encoded_len, Some(encoded))
	}
```

Opinions?

**Comment by lamafab** [comment]
This was discussed in great detail via chat.

TL;DR: If the function call panics, just throw the block away. The Host cannot do anything about it.

![image](https://user-images.githubusercontent.com/42901763/90150915-6fd7ef00-dd86-11ea-913e-18840f549fc8.png)
